// crewai/src/crewai/project/crew_base.ts
import { v4 as uuidv4 } from 'uuid';
import { Agent } from './agent';
import { Task } from './task';
import { Process } from './process';

class CrewBase {
    constructor(agents = [], tasks = [], processes = [], verbose = false) {
        this.id = uuidv4();
        this.agents = agents;
        this.tasks = tasks;
        this.processes = processes;
        this.verbose = verbose;
    }

    addAgent(agent) {
        if (!(agent instanceof Agent)) {
            throw new Error("Only Agent instances can be added to the crew");
        }
        this.agents.push(agent);
    }

    addTask(task) {
        if (!(task instanceof Task)) {
            throw new Error("Only Task instances can be added to the crew");
        }
        this.tasks.push(task);
    }

    addProcess(process) {
        if (!(process instanceof Process)) {
            throw new Error("Only Process instances can be added to the crew");
        }
        this.processes.push(process);
    }

    toJSON() {
        return {
            id: this.id,
            agents: this.agents,
            tasks: this.tasks,
            processes: this.processes,
            verbose: this.verbose
        };
    }
}

export { CrewBase };
