// crewai/src/crewai/utilities/planning_handler.ts
import { v4 as uuidv4 } from 'uuid';

class PlanningHandler {
    constructor(agent, task) {
        this.agent = agent;
        this.task = task;
        this.planId = uuidv4();
    }

    async createPlan() {
        const planPrompt = this.createPlanPrompt();
        const plan = await this.agent.execute(planPrompt);
        return this.parsePlan(plan);
    }

    createPlanPrompt() {
        return `You are ${this.agent.name}, an expert in ${this.agent.role}. Your task is to create a detailed plan to accomplish the following task:

${this.task.description}

Please provide a step-by-step plan, with each step clearly numbered and described. Be specific and thorough in your planning.

Your plan:`;
    }

    parsePlan(plan) {
        const steps = plan.split('\n').filter(step => step.trim() !== '');
        return steps.map((step, index) => ({
            id: `${this.planId}-${index + 1}`,
            description: step.replace(/^\d+\.\s*/, '').trim()
        }));
    }
}

export { PlanningHandler };
